### ISSUE TITLE
Nested clusters

### ISSUE THREAD

**Issue by TkinterinShanghai** [tom-whitehead/hdbscan#4]
Hey! Thanks for implementing this in Rust. I'm grateful for that.

I am having an issue where the clusters are nested. These are the alpha Shapes drawn from the clusters:

<img width="1840" alt="Screenshot 2024-03-19 at 21 37 45" src="https://github.com/tom-whitehead/hdbscan/assets/46348678/9fe69d50-ef8a-44e0-8c05-90618703e645">

As far as I know it should only return the most stable cluster for each instance. I believe the Python implementation handles this differently. Did I misunderstand something?

**Comment by tom-whitehead** [comment]
First issue! Thanks for raising it. As you mention there should only be the one most stable cluster for each data point. 
Are you able to share the data that you are using so that I can debug?

**Comment by TkinterinShanghai** [comment]
The clustering is fantastic and if I had 


to chose a different scale, the larger cluster would have been my go to choice. This is the clustering including the points from two angles:

<img width="1840" alt="Screenshot 2024-03-20 at 13 10 56" src="https://github.com/tom-whitehead/hdbscan/assets/46348678/79d24ef2-07d0-4800-8be8-e0ed318fcafc">
<img width="1840" alt="Screenshot 2024-03-20 at 13 11 06" src="https://github.com/tom-whitehead/hdbscan/assets/46348678/3319aecd-70ed-414d-83e5-26e8c14e4827">

This is the data that I am passing to the clustering function:
[data.json](https://github.com/tom-whitehead/hdbscan/files/14666232/data.json)

This is the function that is using your library:
```rust
use chull::ConvexHullWrapper;
use hdbscan::{Hdbscan, HdbscanHyperParams};
use serde_wasm_bindgen::to_value;
use std::collections::HashMap;
use wasm_bindgen::prelude::*;
use web_sys::console;

#[wasm_bindgen]
pub fn cluster_hdbscan(
    data: Vec<f32>,
    min_cluster_size: usize,
    min_samples: usize,
) -> Result<JsValue, JsValue> {
    console::time_with_label("Total Processing Time");
    let points: Vec<Vec<f32>> = data.chunks(3).map(|chunk| chunk.to_vec()).collect();

    let hyper_params = HdbscanHyperParams::builder()
        .min_cluster_size(min_cluster_size)
        .max_cluster_size(points.len() / 2)
        .min_samples(min_samples)
        .build();

    let clusterer = Hdbscan::new(&points, hyper_params);
    let clustering_result: Vec<i32> = clusterer
        .cluster()
        .map_err(|e| JsValue::from_str(&e.to_string()))?;

    // Group points by cluster label for convex hull calculation
    let mut clusters: HashMap<i32, Vec<Vec<f64>>> = HashMap::new();
    let mut cluster_point_indices: HashMap<i32, Vec<usize>> = HashMap::new();

    for (i, &label) in clustering_result.iter().enumerate() {
        if label >= 0 {
            clusters
                .entry(label)
                .or_insert_with(Vec::new)
                .push(points[i].iter().map(|&x| x as f64).collect());
            cluster_point_indices
                .entry(label)
                .or_insert_with(Vec::new)
                .push(i); // Store the index of each point in its respective cluster
        }
    }

    let hulls_info: Vec<_> = clusters
        .into_iter()
        .map(|(label, cluster_points)| {
            match ConvexHullWrapper::try_new(&cluster_points, None) {
                Ok(hull) => {
                    let (vertices, indices) = hull.vertices_indices();
                    let point_indices = cluster_point_indices.get(&label).cloned().unwrap_or_default();
                    Some((label, vertices, indices, point_indices)) // Include point indices in the output
                },
                Err(e) => {
                    let error_message = format!("Error calculating convex hull: {:?}", e);
                    console::error_1(&JsValue::from_str(&error_message));
                    None
                }
            }
        })
        .filter_map(|x| x)
        .collect();

    console::time_end_with_label("Total Processing Time");
    // Serialize hulls_info, which now includes point indices, to JsValue to return
    to_value(&hulls_info).map_err(|e| JsValue::from_str(&e.to_string()))
}

```

**Comment by TkinterinShanghai** [comment]
Maybe these lines here are not working correctly, I will also debug this in my code

```rust
for (i, &label) in clustering_result.iter().enumerate() {
    if label >= 0 {
```

**Comment by tom-whitehead** [comment]
Thanks for sharing, I'll have a look when I get a chance. This isn't my day job, so it might take a little while, but I will get to it. 

On the scale of the clusters, you could try and influence it with some hyper parameter tuning of `min_cluster_size` and `min_samples`. Otherwise, HDBSCAN will just select the clusters that "live the longest" at all densities.

Out of interest, what is the library that you are using to make the alpha shape plots?

**Comment by tom-whitehead** [comment]
This is definitely a bug. The Python scikit learn version produces the following clusters with same hyper parameters:
![image](https://github.com/tom-whitehead/hdbscan/assets/99190557/58e7168b-f75e-4f0a-bea6-ffdc6718f842)

Compared to my Rust version:
![image](https://github.com/tom-whitehead/hdbscan/assets/99190557/a637ef35-a015-4f96-9853-873cde3b751d)

**Comment by TkinterinShanghai** [comment]
Regarding the time this might take: Thats totally valid, no worries. 
I am using DeckGL for that and React.
Let me know if I can help you debug this. It would be great if we could get this to work. I really want to use your implementation in my thesis, so that I can adjust the cluster size interactively in the browser

**Comment by tom-whitehead** [comment]
Hey @TkinterinShanghai I'm just publishing a fix for this now. Actually I found another bug during the debugging of this one, so killed two birds with one stone. The new version will be `0.4.1`. I've tested it on your data and it seems to be working well, but please give it a go and check it's working as expected.

I appreciate also that it isn't exactly performant on a larger dataset. My next improvement will be to optimise the step that calculates of the minimum spanning tree of the data points, as this seems to be the bottle neck.

**Comment by TkinterinShanghai** [comment]
Hey! Awesome! I'll check it out ASAP. A performance improvement would be nice as well. Currently it takes about 8 seconds to calculate the clusters on my website. Not bad, but a reduction would definitely be appreciated :).

**Comment by TkinterinShanghai** [comment]
Works like a charm! You can check it out here: tobias-wetzel.com/animation. It takes a while until the clusters appear. I might be able to add a loading bar. The performance optimization would be awesome. It doesn't have high priority for me though. If I have some time once I am done with everythingelse, I might be able to contribute

**Comment by tom-whitehead** [comment]
Hey @TkinterinShanghai , just letting you know that I've published a performance optimisation just now (`0.5.0`). On my machine it shaves 1 second plus off the clustering of your dataset!

**Comment by TkinterinShanghai** [comment]
Super awesome! On the Web it reduced the time by two seconds even! Thanks for that :)

Thats the old one
<img width="1840" alt="Screenshot 2024-04-04 at 15 38 50" src="https://github.com/tom-whitehead/hdbscan/assets/46348678/c90d76f6-3fce-412b-b958-a457083c84a1">
And thats the new one
<img width="1840" alt="Screenshot 2024-04-04 at 15 42 41" src="https://github.com/tom-whitehead/hdbscan/assets/46348678/6d2038e8-55df-4b78-8f28-198f1fbbb7db">
